dask/task_spec.py

```diff
@@ -0,0 +1,912 @@
+                del sub_dsk
+
+            new_args_l = map(partial(inner, all_keys=all_keys_inner), args)
+
+            func2 = func.inline(dict(zip(subgraph.inkeys, new_args_l)))
+            dct = {k: Alias(k) for k in func2.dependencies}
+            return Task(key=k, func=func2, args=[dct])
+        else:
+            new_args = map(inner_args, args)
+            if only_refs:
+                # this can't be a literal since the literal wouldn't execute
+                # new_args in case there were runnable tasks or other literals
+                # See test_convert_task_only_ref
+                return Task(k, identity, (func, *new_args))
+            return Task(key=k, func=func, args=tuple(new_args))
+    try:
+        if isinstance(arg, (bytes, int, float, str, tuple)) and arg in all_keys:
+            if arg in cache:
+                return cache[arg]
+            cache[arg] = rv = Alias(arg)
+            return rv
+    except TypeError:
+        # Unhashable
+        pass
+
+    if isinstance(arg, (list, tuple, set, frozenset)):
+        if is_namedtuple_instance(arg):
+            rv = _wrap_namedtuple_task(k, arg, inner_args)
+        else:
+            rv = SequenceOfTasks(k, map(inner_args, arg), typ=type(arg))
+        return rv.propagate_literal()
+    elif isinstance(arg, dict):
+        new_dsk = convert_old_style_dsk(
+            arg, all_keys=all_keys, only_refs=True, cache=cache
+        )
+
+        return DictOfTasks(None, new_dsk).propagate_literal()
+
+    elif isinstance(arg, WrappedKey):
+        return Alias(arg.key)
+    else:
+        return arg
+
+
+def convert_old_style_dsk(
+    dsk: Mapping,
+    all_keys: Container | None = None,
+    only_refs: bool = False,
+    cache: MutableMapping | None = None,
+):
+    if not all_keys:
+        all_keys = set(dsk)
+    new_dsk = {}
+    if cache is None:
+        cache = {}
+    for k, arg in dsk.items():
+        t = convert_old_style_task(k, arg, all_keys, only_refs, cache=cache)
+        if not only_refs and isinstance(t, Alias) and t.key == k:
+            continue
+        new_dsk[k] = t
+    new_dsk2 = {
+        k: v if isinstance(v, BaseTask) else LiteralTask(k, v)
+        for k, v in new_dsk.items()
+    }
+    return new_dsk2
+
+
+def resolve_aliases(dsk: dict, keys: set, dependents: dict) -> dict:
+    """Remove trivial sequential alias chains
+
+    Example:
+
+        dsk = {'x': 1, 'y': Alias('x'), 'z': Alias('y')}
+
+        resolve_aliases(dsk, {'z'}, {'x': {'y'}, 'y': {'z'}}) == {'z': 1}
+
+    """
+    if not keys:
+        raise ValueError("No keys provided")
+    dsk = dict(dsk)
+    work = list(keys)
+    seen = set()
+    while work:
+        k = work.pop()
+        if k in seen or k not in dsk:
+            continue
+        seen.add(k)
+        t = dsk[k]
+        if (
+            isinstance(t, Alias)
+            and t.key not in keys
+            and t.key != k
```

Comment:
What would actually happen in this self-referential case?

Reply:
at the very least dask.order raises a "cycles found" exception because strictly speaking a graph with this kind of self reference is not a DAG.

We had similar logic in the distributed scheduler before so without this kind of filtering, nothing will work. It's just a question about where we do this